pulser-core/pulser/channels/base_channel.py

```diff
@@ -391,22 +400,41 @@ def modulate(
             return input_samples
         else:
             mod_bandwidth = self.mod_bandwidth
-            rise_time = self.rise_time
+            mod_padding = self._modulation_padding
 
-        # The cutoff frequency (fc) and the modulation transfer function
-        # are defined in https://tinyurl.com/bdeumc8k
-        fc = mod_bandwidth * 1e-3 / np.sqrt(np.log(2))
         if keep_ends:
-            samples = np.pad(input_samples, 2 * rise_time, mode="edge")
+            samples = np.pad(
+                input_samples, mod_padding + self.rise_time, mode="edge"
```

Comment:
I'm not sure about spectral leakage. The input samples are padded because the output signal is longer by about `rise_time` on each side. 
By default, we pad the signal with zeros, but when `keep_ends=True`, we want to make it seem like the signal started out from the first value and at the end stays at the last value, so we pad with `mode="edge"` instead. Plus, we use a longer padding time that we then truncate such that the modulation doesn't show the signal ramping to the starting and ending values at the edges.